pallets/pallet-asset-swap/src/lib.rs

```diff
@@ -0,0 +1,534 @@
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		const POOL_ADDRESS_GENERATION_ENTROPY: [u8; 8];
+
+		type AccountIdConverter: TryConvert<Self::AccountId, Junction>;
+		type AssetTransactor: TransactAsset;
+		type FeeOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+		type LocalCurrency: MutateFungible<Self::AccountId>;
+		type PauseOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		type SubmitterOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Self::AccountId>;
+		type SwapOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+		type XcmRouter: SendXcm;
+	}
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		SwapPairCreated {
+			remote_asset_id: VersionedAssetId,
+			remote_asset_reserve_location: VersionedMultiLocation,
+			remote_xcm_fee: Box<VersionedMultiAsset>,
+			total_issuance: u128,
+			circulating_supply: u128,
+			pool_account: T::AccountId,
+		},
+		SwapPairRemoved {
+			remote_asset_id: VersionedAssetId,
+		},
+		SwapPairResumed {
+			remote_asset_id: VersionedAssetId,
+		},
+		SwapPairPaused {
+			remote_asset_id: VersionedAssetId,
+		},
+		SwapPairFeeUpdated {
+			old: VersionedMultiAsset,
+			new: VersionedMultiAsset,
+		},
+		SwapExecuted {
+			from: T::AccountId,
+			to: VersionedMultiLocation,
+			amount: LocalCurrencyBalanceOf<T>,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		InvalidInput,
+		LocalPoolBalance,
+		Liquidity,
+		PoolInitialLiquidityRequirement,
+		SwapPairAlreadyExisting,
+		SwapPairNotEnabled,
+		SwapPairNotFound,
+		UserSwapBalance,
+		UserXcmBalance,
+		Xcm,
+		Internal,
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn swap_pair)]
+	pub(crate) type SwapPair<T> = StorageValue<_, SwapPairInfoOf<T>, OptionQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		LocalCurrencyBalanceOf<T>: Into<u128>,
+	{
+		#[pallet::call_index(0)]
+		#[pallet::weight(u64::MAX)]
+		pub fn set_swap_pair(
+			origin: OriginFor<T>,
+			reserve_location: Box<VersionedMultiLocation>,
+			remote_asset_id: Box<VersionedAssetId>,
+			remote_fee: Box<VersionedMultiAsset>,
+			total_issuance: u128,
+			circulating_supply: u128,
+		) -> DispatchResult {
+			T::SwapOrigin::ensure_origin(origin)?;
+
+			// 1. Verify swap pair has not already been set.
+			ensure!(!SwapPair::<T>::exists(), Error::<T>::SwapPairAlreadyExisting);
+			// 2. Verify that total issuance >= circulating supply and take the difference
+			//    as the amount of assets we control on destination.
+			let locked_supply = total_issuance
+				.checked_sub(circulating_supply)
+				.ok_or(Error::<T>::InvalidInput)?;
+			// 3. Verify the pool account does not exist and has enough local assets to
```

Comment:
The comment is wrong, yes, since we need to check that the account has enough funds. But this is a pre-condition: a pool for an asset cannot be created if its to-be pool account does not have enough assets to cover all the swaps, so it needs to be pre-funded with a regular transfer by whoever.